FIX stratification in StratifiedGroupKFold when shuffle=True

pfolch opened the pull request:
#### Reference Issues/PRs
Fixes #32478 
cc: @ogrisel 

#### What does this implement/fix? Explain your changes.
When `StratifiedGroupKFold` is instantiated with `shuffle=True`, the implementation shuffles `y_counts_per_group` in place, but does **not** update the internal `groups_inv` mapping that encodes each sample’s group index. This desynchronizes the “row i ↔ group i” invariant. The greedy assignment then operates on permuted rows, while the final test indices are gathered using the unpermuted `groups_inv`, yielding test folds that correspond to **incorrect (effectively random) groups**. In practice, this makes the behavior equivalent to `GroupKFold` with a random group assignment rather than a stratified grouping.

**Change**: when `shuffle=True`, apply a single permutation to `y_counts_per_group` **and** remap `groups_inv` by the inverse permutation. This keeps the internal encoding aligned so that each row in `y_counts_per_group` still refers to the same group index used to build the test folds. The stable variance-based sort continues to use the shuffled order only as a tie-breaker for groups with identical class-variance.

No API changes; negligible performance impact. Reproducibility with an integer `random_state` is preserved.

#### Any other comments?
**Test added:** `test_stratified_group_kfold_shuffle_preserves_stratification`: non-regression test that, across many random seeds, checks:
-   no group appears on both sides of any split,
-   train/test class **distribution** match the global dataset distribution,
-   fold sizes remain balanced (peak-to-peak ≤ 1).
This reproduces the scenario from the issue and fails on main prior to this fix.

**Changelog:** a short entry was added to document the bug fix in `StratifiedGroupKFold` when `shuffle=True`.

pfolch committed:
TEST: Added non-regression for issue #32478

pfolch committed:
FIX: StratifiedGroupKFold shuffle group indices

pfolch committed:
DOC: Added changelog for issue #32478

pfolch committed:
FIX: Rename changelog entry to match PR id instead of issue id

ogrisel reviewed (approved):
LGTM besides the following nitpicks:

pfolch committed:
refactor: Rename variable to make it more explicit

pfolch committed:
Update doc/whats_new/upcoming_changes/sklearn.model_selection/32540.fix.rst

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

pfolch committed:
Merge branch 'main' into fix-StratifiedGroupKFold-shuffle-indices

pfolch commented:
Thanks a lot for the quick review and helpful suggestions, @ogrisel.
I’ve renamed the variable to `expected_class_ratios` and accepted the changelog edit. I also updated the PR, and CI is green.

lucyleeow commented on doc/whats_new/upcoming_changes/sklearn.model_selection/32540.fix.rst:
```diff
@@ -0,0 +1,3 @@
+- Fix shuffle behaviour in :class:`model_selection.StratifiedGroupKFold`. Now
+  stratification among folds is also preserved when `shuffle=True`.
+  By :user:`Pau Folch <pfolch>`
```
```suggestion
  By :user:`Pau Folch <pfolch>`.
```
nit

lucyleeow commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -724,6 +724,37 @@ def test_stratified_group_kfold_homogeneous_groups(y, groups, expected):
+def test_stratified_group_kfold_shuffle_preserves_stratification():
+    # Check StratifiedGroupKFold with shuffle=True preserves stratification:
+    # shuffling only affects tie-breaking among groups with identical
+    # standard deviation of class distribution
```
```suggestion
    # standard deviation of class distribution (see #32478)
```

lucyleeow reviewed (approved):
Thanks @pfolch ! LGTM!

Small doc nits amended only.

lucyleeow committed:
Update doc/whats_new/upcoming_changes/sklearn.model_selection/32540.fix.rst

lucyleeow committed:
Update sklearn/model_selection/tests/test_split.py
